cpp/dynamic_community_detection_module/algorithm/dynamic_community_detection.cpp

```diff
@@ -0,0 +1,370 @@
+    std::unique_ptr<mg_graph::Graph<>>& graph,
+    std::unordered_set<std::uint64_t> changed_nodes,
+    std::unordered_set<std::uint64_t> to_delete, bool persist) {
+  this->graph = std::move(graph);
+
+  bool incremental = changed_nodes.size() >= 1 ? true : false;
+  if (incremental) {
+    RemoveDeletedNodes(to_delete);
+    for (const auto node_id : changed_nodes) {
+      SetStructures(node_id);
+    }
+  } else {
+    label_Ps.clear();
+    sum_w.clear();
+    times_updated.clear();
+    for (const auto node_id : NodesMemgraphIDs()) {
+      SetStructures(node_id);
+    }
+  }
+
+  for (std::uint64_t i = 0; i < this->max_iterations; i++) {
+    auto [none_updated, most_updates] =
+        Iteration(incremental, changed_nodes, to_delete);
+    if (none_updated || most_updates > this->max_updates) break;
+  }
+
+  if (persist) this->calculated = true;
+  ResetTimesUpdated();
+
+  return AllLabels();
+}
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::GetLabels(
+    std::unique_ptr<mg_graph::Graph<>>& graph) {
+  std::unordered_map<std::uint64_t, std::int64_t> labels;
+  if (!calculated) return CalculateLabels(graph, {}, {}, false);
+
+  return AllLabels();
+}
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::SetLabels(
+    std::unique_ptr<mg_graph::Graph<>>& graph, bool is_directed,
+    bool is_weighted, double similarity_threshold, double exponent,
+    double min_value, std::string weight_property, double w_selfloop,
+    std::uint64_t max_iterations, std::uint64_t max_updates) {
+  this->is_directed = is_directed;
+  this->is_weighted = is_weighted;
+  this->similarity_threshold = similarity_threshold;
+  this->exponent = exponent;
+  this->min_value = min_value;
+  this->weight_property = weight_property;
+  this->w_selfloop = w_selfloop;
+  this->max_iterations = max_iterations;
+  this->max_updates = max_updates;
+
+  return CalculateLabels(graph);
+}
+
+std::unordered_map<std::uint64_t, std::int64_t> LabelRankT::UpdateLabels(
+    std::unique_ptr<mg_graph::Graph<>>& graph,
+    std::vector<std::uint64_t> modified_nodes,
+    std::vector<std::pair<std::uint64_t, std::uint64_t>> modified_edges,
+    std::vector<std::uint64_t> deleted_nodes,
+    std::vector<std::pair<std::uint64_t, std::uint64_t>> deleted_edges) {
+  if (!calculated) return CalculateLabels(graph, {}, {}, false);
+
+  std::unordered_set<std::uint64_t> changed_nodes(modified_nodes.begin(),
+                                                  modified_nodes.end());
+  std::unordered_set<std::uint64_t> to_delete(deleted_nodes.begin(),
+                                              deleted_nodes.end());
+
+  for (const auto edge : modified_edges) {
+    changed_nodes.insert(edge.first);
+    changed_nodes.insert(edge.second);
+  }
+
+  for (const auto edge : deleted_edges) {
+    if (to_delete.count(edge.first) == 0) changed_nodes.insert(edge.first);
+    if (to_delete.count(edge.second) == 0) changed_nodes.insert(edge.second);
```

Comment:
```suggestion
    changed_nodes.insert(edge.first);
    changed_nodes.insert(edge.second);
```

The insert will handle the duplications without any additional counts.

Reply:
If we delete both a node and its incident edge, the deleted node would be added to `changed_nodes` per your suggestion, even though it should belong to `deleted_nodes` and nothing else.
This ensues in a crash - changed nodes have their communities updated and this cannot be done for a deleted node.

Reply:
Oh sorry, I missed the part that the count is called on a different container, sorry.